jupyter-client.el

```diff
@@ -946,24 +946,32 @@ client local variable.
 
 Methods that extend this generic function should
 `cl-call-next-method' as a last step."
-  (cl-check-type jupyter-current-client jupyter-kernel-client
-                 "Need a client to read an expression")
   (let* ((client jupyter-current-client)
          (jupyter--read-expression-history
           (jupyter-get client 'jupyter-eval-expression-history)))
-    (minibuffer-with-setup-hook
-        (lambda ()
-          (setq jupyter-current-client client)
-          (add-hook 'completion-at-point-functions
-                    'jupyter-completion-at-point nil t)
-          (add-hook 'minibuffer-exit-hook
-                    'jupyter--teardown-minibuffer nil t))
-      (prog1 (read-from-minibuffer
-              (format "Eval (%s): " (jupyter-kernel-language client))
-              nil read-expression-map
-              nil 'jupyter--read-expression-history)
-        (jupyter-set client 'jupyter-eval-expression-history
-                     jupyter--read-expression-history)))))
+    (prog1
+        (jupyter--read-with-completion
+         client "Eval (%s): " jupyter--read-expression-history)
+      (jupyter-set client 'jupyter-eval-expression-history
+                   jupyter--read-expression-history))))
+
+(defun jupyter--read-with-completion (client prompt &optional history)
+  "Read an expression using CLIENT for completion.
+The expression is read from the minibuffer with PROMPT and expression
+history HISTORY."
+  (cl-check-type client jupyter-kernel-client
+                 "Need a client to read an expression")
+  (minibuffer-with-setup-hook
+      (lambda ()
+        (setq jupyter-current-client client)
+        (add-hook 'completion-at-point-functions
```

Comment:
This should probably use `completing-read` instead. Any idea why it was written this way instead?

Reply:
I copied it over from `read--expression` called by `eval-expression` (`M-:`). My intention with `jupyter-read-expression` is to parallel `read--expression`. Also `completing-read` is more for completing collections right? `jupyter-read-expression` is supposed to read a line of code for evaluation from the minibuffer while also allowing completion from the kernel within the line.

Although the documentation for `jupyter-read-expression` does say `for completion` instead of something more approriate like `for evalution` so its purpose isn't that clear.

Reply:
Hm. That makes sense. So for this new purpose I'm putting this code to, using `completing-read` probably makes more sense, but it doesn't make sense in the context of `jupyter-read-expression` which is likely to complete multiple symbols. I think this is an acceptable hack though until I finish a `completing-read` based implementation?

Reply:
In either case, I think it probably makes more sense to just refactor `jupyter-read-expression` to take optional parameters rather than define an internal `jupyter--read-with-completion`. :/